H003 schematic support

jplot opened the pull request:
Testing overview:

| Order Type | H003 | H004 | H005 | Description |
|-------------|------|------|------|-------------|
| AZV | ❌ | ✅ | ❌ | |
| B2B | ❌ | ✅ | ❌ | |
| BTD | ❌ | ❌ | ✅ | |
| BTU | ❌ | ❌ | ✅ | |
| BKA | ❌ | ✅ | ❌ | |
| C2S | ❌ | ✅ | ❌ | |
| C5N | ❌ | ✅ | ❌ | |
| C52 | ❌ | ✅ | ✅ | |
| C53 | ❌ | ✅ | ✅ | |
| C54 | ❌ | ✅ | ✅ | |
| CCS | ❌ | ✅ | ❌ | |
| CCT | ❌ | ✅ | ✅ | Used to initiate the credit transfer per Single Euro Payments Area (SEPA) specification set by the European Payment Council or Die Deutsche Kreditwirtschaft (DK). |
| CD1 | ❌ | ✅ | ❌ | Uploads a SEPA Direct Debit document of type COR1 |
| CDB | ❌ | ✅ | ✅ | Uploads a SEPA Direct Debit document of type B2B |
| CDD | ❌ | ✅ | ✅ | Uploads a SEPA Direct Debit document of type CORE |
| CDS | ❌ | ✅ | ❌ | |
| CDZ | ❌ | ✅ | ❌ | |
| CIP | ❌ | ✅ | ✅ | |
| CRZ | ❌ | ✅ | ❌ | |
| FDL | ✅ | ✅ | ✅ | |
| FUL | ❌ | ❌ | ❌ | |
| H3K | ❌ | ❌ | ❌ | |
| HAA | ✅ | ✅ | ✅ | Download retrievable order types |
| HAC | ✅ | ✅ | ✅ | Used to download customer acknowledgment (XML format) |
| HEV | ✅ | ✅ | ✅ | Download of supported EBICS versions |
| HIA | ✅ | ✅ | ✅ | Used to transmit user public certificates for identification, authentication and encryption within the framework of subscriber initialization |
| HKD | ✅ | ✅ | ✅ | Retrieve customer's customer and subscriber information |
| HPB | ✅ | ✅ | ✅ | Fetch your bank’s public keys |
| HPD | ✅ | ✅ | ✅ | Download bank parameters |
| HTD | ✅ | ✅ | ✅ | Retrieve subscriber’s customer and subscriber information |
| INI | ✅ | ✅ | ✅ | Used in subscriber initialization. Sends the bank-technical public certificate of the user to the EBICS Banking Server |
| PTK | ✅ | ✅ | ✅ | Customer usage report in text format |
| SPR | ❌ | ❌ | ❌ | |
| STA | ❌ | ✅ | ✅ | Statements in MT940 format |
| VMK | ❌ | ✅ | ✅ | Customer usage report in XML format |
| WSS | ❌ | ✅ | ❌ | |
| XCT | ❌ | ✅ | ❌ | |
| XDS | ❌ | ✅ | ❌ | |
| XE2 | ❌ | ✅ | ✅ | |
| XE3 | ❌ | ✅ | ✅ | |
| XEK | ❌ | ❌ | ✅ | |
| Z52 | ❌ | ✅ | ✅ | |
| Z53 | ❌ | ✅ | ✅ | |
| Z54 | ❌ | ✅ | ✅ | |
| ZSR | ❌ | ❌ | ✅ | |
| YCT | ❌ | ❌ | ✅ | |

jplot committed:
Adds h003 schematic support

jplot committed:
Adds a005 signature support

jplot committed:
Adds h005 schematic support

jplot commented:
Hi @tobischo,
You can start reviewing the code.
I'm waiting to finalize tests with our financial partner before putting the PR on ready.

tobischo commented:
I am really excited to see that you also added EBICS 3 🤩 

I had that on my list as a potential followup after your changes.
Will review as soon as I have time and also start testing whether the 2.5 cases still work that I am responsible
I might have access on a test API to 3.0 as well 🤔

tobischo commented on README.md:
```diff
@@ -6,7 +6,7 @@
 Communication Standard).
 
-It supports EBICS 2.5.
+It supports EBICS 2.4, 2.5 and 3.0.
```
I think the readme could reflect which one is the default when initializing a new or existing client and how to work with another version

tobischo commented on lib/epics/header_request.rb:
```diff
@@ -57,4 +60,13 @@ def build_attributes(xml, attributes)
     end
   end
+
+  def b36encode(number)
```
I am wondering: was `number.to_s(36).upcase` not a viable option?

tobischo commented on spec/key_spec.rb:
```diff
@@ -1,4 +1,4 @@
-RSpec.describe Epics::Key do
+RSpec.describe Epics::SignatureAlgorithm::RsaPss do
```
It might make sense to change the filename and location accordingly to the class it tests

tobischo reviewed (changes requested):
Overall this looks great.

I haven't tested it yet against an EBICS API, so I cannot speak to whether it is still working (yet).

I think it is missing some test cases, e.g. around the signature algorithm handling and using a different EBICS version

jplot committed:
Fixes after code review

frantisekrokusek commented:
@jplot do you need help finalizing this ?

jplot commented:
> @jplot do you need help finalizing this ?

I don't have a sandbox environment to test end-to-end operation.

tobischo commented:
I haven't checked if it supports EBICS 2.4, but https://iso20022test.credit-suisse.com/login provides a testing environment for communication through EBICS

janz93 commented:
Sadly the credit-suisse test ox does not work for ebics 3.0
>Es gibt keine Testumgebung; insbesondere eignet sich der
EBICS-3.0-Zugang der Credit Suisse ISO 20022 Test
Platform nicht zum Testen von Direct Link

https://www.credit-suisse.com/media/assets/private-banking/docs/ch/unternehmen/unternehmen-unternehmer/umstellung-von-ebics-25-auf-30-de.pdf

jplot committed:
Fixes order id for ebics 2.4

jplot committed:
Fixes letter renderer

jplot committed:
Refacts keys management

jplot committed:
Adds schematics for version H003

jplot committed:
Adds H003 schema validity tests

jplot committed:
Adds schematics for version H005

jplot committed:
Fixes failure message for be_a_valid_ebics_doc matcher

jplot commented:
@tobischo I have reallocated my work time to this project full-time, with the aim of integrating the features related to versions 'H003' and 'H005'.
To ensure compatibility with version 'H005', it was imperative to refactor the way we generate XML.

jplot committed:
Adds new xml schema generator

tobischo commented:
Amazing 😃
Let me know when you think it is testable

frantisekrokusek commented:
@jplot We have some mandates with H003 servers so we can also do the testing part for you 😉

jplot commented:
@frantisekrokusek If you have the possibility of writing more complete tests on the main branch, this would allow me to guarantee the integrity of the new system.

jplot committed:
Uses new xml schema generator

jplot commented:
For version H003, I can only confirm that FDL, HAA, HAC, HEV, HIA, HKD, HPB, HPD, HTD, INI, PTK are supported.

jplot committed:
Adds early support for version H005

jplot commented:
![StackEdit-06-18-2025_04_05_PM](https://github.com/user-attachments/assets/50429f74-3767-480b-b3d2-a45ff83aead1)

jplot commented:
Key Technical Points for H003 and H005 Versions

- AuthSignature: We've observed that the AuthSignature function isn't working correctly on version H003. An investigation is needed to resolve this issue.
- X.509 Certificate Generation: The implementation of X.509 certificate generation will become mandatory for version H005. Therefore, we must ensure this functionality is properly integrated.
- Signature Functions (EMSAs): For version H005, the emsaPssEncode (reference A006) and emsaPkcs1V15Encode (reference A005) signature functions are missing. Their development is crucial for compliance and security.

jplot commented:
I'll take some time this week to update the branch, look into leveraging the X509 Certificate support, and wrap up the remaining items.

jplot committed:
Migrate Z01 order type to factories/builders architecture

Replaces the legacy header_request.build approach with the versioned
factory pattern (V25: build_download_request, V3: BTD with PSR/pain.002).
Adds z01 to ORDER_TYPES registry and updates spec with shared examples.

jplot committed:
Migrate FUL order type to factories/builders architecture

Replaces the legacy header_request.build approach with the versioned
factory pattern. Adds add_ful_order_params to OrderDetailsBuilder for
FileFormat support, and makes build_upload_request accept a block for
custom order params. Updates spec with shared examples.

jplot committed:
Fix version support correctness for Z01 and FUL

- Fix V2 factory: %w[btd btu] instead of ['btd btu'] so create_btd
  and create_btu properly raise VersionSupportError
- V3 Z01: add service_option CH003GEN per SIX Group spec
- V3 FUL: raise VersionSupportError since FUL doesn't exist in H005

jplot committed:
Add comprehensive H004 structural tests for all order types and phases

Add shared examples for receipt, transfer, FDLOrderParams, and
FULOrderParams phases. Apply H004 structural validation to all 30
order types covering init, transfer, and receipt phases as applicable.

jplot committed:
Parameterize structural shared examples for H003/H004 and add H003 tests

Rename 'a valid H004 ...' shared examples to 'a valid ebicsRequest ...'
with an ebics_version parameter (default 'H004') to support both H003
and H004 structural validation from the same test definitions.

Add H003 structural tests for the 8 order types that already support it:
HKD, HPD, HTD, PTK (init), HAA, HAC (init + receipt), STA, FDL (receipt).

jplot committed:
Add H003 structural tests for all order types

Add pending H003 init blocks for downloads (BKA, C52-C54, C5N, VMK,
WSS, Z01, Z52-Z54, CDZ, CRZ, STA) and uploads (CCT, B2B, CD1, CDB,
CDD, CIP, XE2, XE3, AZV, C2S, CCS, FUL). Add active H003 transfer
blocks for all uploads since transfer XML is version-agnostic.

Fix FDL: remove incorrect H003 pending markers since FDL already
supports H003 through the factories/builders architecture.

jplot committed:
Add H003 download and upload support for all order types

Move build_download_request, build_upload_request, and all 27 create_xxx
methods from V25 up to V2, making them available to both V24 (H003) and
V25 (H004). Fix V24 add_order_type to support OZHNN for uploads with
electronic signature. Remove duplicate V24 initialize.

V24 and V25 now differ only by namespace (H003/H004) and OrderID
(mandatory in H003, absent in H004).

jplot committed:
Consolidate X509 certificate handling on Epics::Crypt::X509

Remove Epics::X509Certificate in favor of Epics::Crypt::X509 which now
provides data, fingerprint and to_pem. Certificates are assigned directly
on Signature objects via the keyring instead of being stored separately
on the client.

Remove x_509_certificates_content, x_509_certificate(type) and
x_509_certificate_hash(type) from Client. Remove dead x509_data_xml
from GenericRequest. Update OrderDataHandler, LetterRenderer and
ini_with_certs template to use certificates from signatures.

jplot commented:
Would appreciate if folks could test this branch against their EBICS connections to confirm H003 and H004 both work as expected. Any feedback welcome!

jplot committed:
Activate H005 support: certificates, CryptService.sign, BTD/BTU mappings

- Add auto-generated X.509 certificates for H005 test clients
- Implement CryptService.sign() with RSA-PSS (A006) and PKCS1v15 (A005)
- Add BTD mappings for BKA, C5N, CDZ, CRZ downloads in V3 factory
- Add BTU mappings for AZV, B2B, CCS, CDS, C2S uploads in V3 factory
- Add VersionSupportError for FDL, CD1, WSS, XDS in H005
- Fix missing AuthSignature in V3 create_btd
- Add H005 structural shared examples (header, download, upload, receipt, transfer)
- Remove all H005 pending tests and replace with active structural tests
- Override AZV#to_xml to use create_azv (separate from CD1 in H005)

jplot committed:
Add unit tests for CryptService and DigestResolver V2/V3

Cover all CryptService public methods (sign A005/A006, encrypt,
AES round-trip, transaction key encryption, hash, digest calculation,
certificate fingerprint) and DigestResolver behavior differences
between V2 (key-based) and V3 (certificate-based).

jplot committed:
Add unit tests for all builder classes

Test XmlBuilder (V24/V25/V3 namespaces and container types),
OrderDetailsBuilder (V2 OrderType/OrderAttribute vs V3 AdminOrderType/
BTDOrderParams/BTUOrderParams), StaticBuilder, MutableBuilder,
DataTransferBuilder (V2 no-op vs V3 DataDigest/AdditionalOrderInfo),
DataEncryptionInfoBuilder, HeaderBuilder, BodyBuilder, and
TransferReceiptBuilder.

jplot committed:
Add unit tests for all handler classes

Test UserSignatureHandler V2 (S001) vs V3 (S002), OrderDataHandler
V24/V25 (PubKeyValue with RSA material) vs V3 (X509Data, no PubKeyValue),
and AuthSignatureHandler (digest, signature verification, C14N) across
all three EBICS versions.

jplot committed:
Add comprehensive unit tests for RequestFactory V24/V25/V3

Test base factory methods (HEV, INI, HIA, HPB, transfers, admin orders),
V2 download/upload patterns with OrderType/OrderAttribute/OrderID,
V3 BTD/BTU patterns with AdminOrderType and service descriptors,
and VersionSupportError boundaries between protocol versions.

jplot committed:
Add unit tests for Keyring, key serialization, and Client.setup

Test version validation, key slot accessors, key serialization
round-trip (dump/reload with AES-256-CBC encryption), Client.setup
key generation (A005/A006, X002, E002), next_order_id overflow,
and urn_schema per protocol version.

jplot committed:
Remove dead Signer class

Signer was only used by the XMLSIG Faraday middleware which has been
replaced by AuthSignatureHandler signing directly during request
construction in the factories.

jplot committed:
Refactor XCT to use factory pattern and remove dead HeaderRequest

XCT now delegates to request_factory.create_azv like all other order
types instead of using the legacy HeaderRequest monolithic builder.
HeaderRequest has no remaining consumers and is removed along with
its require and Client#header_request accessor.

jplot committed:
Fix AZV to inherit from GenericUploadRequest instead of CD1

AZV overrides to_xml completely and uses create_azv (not create_cd1),
so inheriting from CD1 was misleading. AZV and CD1 are unrelated
order types (international payment vs SEPA COR1 direct debit).

janz93 commented:
@jplot nice work!
I'll test it within the next two days and provide you some feedback :)

jplot committed:
Add specs for CDS and XCT order types

CDS: SEPA Direct Debit without ES (DZHNN), BTU with SDD/BIL/pain.008
XCT: International payment alias for AZV (CD1 in V2, XCT/dtazv in V3)

jplot committed:
Add missing scope assertions in H005 specs for XCT and CDD

XCT: add scope 'DE' assertion (matching AZV which already had it)
CDD: add scope 'GLB' assertion for SEPA Direct Debit Core

jplot committed:
Remove unused password attribute from Keyring

The passphrase for key encryption/decryption is stored on Client,
not on Keyring. This attribute was never used anywhere in production.

jplot committed:
Add xsi:schemaLocation to all EBICS XML requests

Include xmlns:xsi namespace declaration and xsi:schemaLocation attribute
on all H003/H004/H005 request root elements (ebicsRequest,
ebicsUnsecuredRequest, ebicsNoPubKeyDigestsRequest, ebicsUnsignedRequest).

Each version points to its official XSD:
- H003: http://www.ebics.org/H003/ebics_H003.xsd
- H004: http://www.ebics.org/H004/ebics_H004.xsd
- H005: http://www.ebics.org/H005/ebics_H005.xsd

Update fixtures and auth signature test values to reflect the new
canonical XML content (C14N inclusive includes ancestor namespaces).

jplot committed:
Add EBICS documentation references

Comprehensive reference file with links to official EBICS specifications,
XSD schemas, implementation guides, BTF/BTD/BTU documentation, and
country-specific resources (France, Switzerland, Germany).

jplot committed:
Merge INI letter templates into a single conditional template

The ini.erb template now handles both certificate and RSA key display
using conditionals, eliminating the need for a separate ini_with_certs.erb.

janz93 commented on lib/epics/letter_renderer.rb:
```diff
@@ -13,30 +19,19 @@ def translate(key, **options)
   alias t translate
 
-  def_delegators :@client, :host_id, :user_id, :partner_id, :a, :x, :e
+  def_delegators @digest_resolver, :confirm_digest
```
I found a typo
```suggestion
  def_delegators :@digest_resolver, :confirm_digest
```

janz93 committed:
Update lib/epics/letter_renderer.rb

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/railslove/epics?pullRequest=154) <br/>All committers have signed the CLA.

janz93 committed:
fixc: servial issue

janz93 committed:
delete me vend

janz93 committed:
style: run linter

janz93 committed:
fix: resolve authentication signature verification with EBICS 3.0 banks

- Canonicalize SignedInfo in the context of the complete EBICS request document rather than a separate XML builder context
- This ensures namespace inheritance matches what banks compute during signature validation
- Fixes interoperability issues where bank servers reject requests due to digest mismatches

janz93 committed:
refactor: use standard error

use standard error in favor of `nil` to not catch ruby default errors.
This helps for the debugging of system errors

janz93 committed:
refactor: remove unneeded `self`

the self prefix is not needed at this method since `setup` creates a instance of the class.

janz93 committed:
style; run linter

janz93 committed:
feat: support X.509 certificates in HPB for EBICS 3.0 compatibility

EBICS 3.0 (H005) uses X.509 certificates instead of raw RSA keys in HPB
responses. This change enables the client to work with banks using the
newer protocol format.

janz93 committed:
style: run linter

janz93 committed:
fix: support bank servers that provide raw RSA keys without X.509 certificates

Some EBICS 3.0 bank servers (e.g., ZKB test platform) only return raw
RSA public keys (PubKeyValue) in HPB responses instead of X.509
certificates.
This aligns V3 behavior with V2, which already handles both cases
correctly by falling back to key-based digest calculation.

janz93 committed:
fix: remove erroneous 09-prefix exclusion from business_error? detection

business_error? was filtering out all return codes starting with "09",
but nearly every EBICS business error (091116, 091001, 090003, etc.)
uses that prefix. This caused business errors to pass silently through
the ParseEbics middleware without raising Epics::Error::BusinessError.

janz93 committed:
style: run linter

janz93 committed:
feat: persist X.509 certificates alongside bank encryption keys

Bank certificates obtained during HPB initialization are required for EBICS 3.0
signature verification. Previously they were stored only in memory, requiring
HPB re-initialization on every process restart. Certificates are now encrypted
and stored in the keyring file to enable offline operation.

janz93 committed:
style: run linter

janz93 committed:
feat: use certificate fingerprints for EBICS 3.0 signature digests

EBICS 3.0 mandates certificate-based digests per AEGIS security requirements.
Without persisted certificates, the client falls back to key-based digests,
which degrades security posture and may cause interoperability issues with
strict EBICS 3.0 servers.

janz93 committed:
fix: canonicalize all matching XPath nodes instead of first only

The at_xpath helper returns only the first match, causing signature validation
failures when multiple authenticated elements exist. Using xpath with map
ensures all matching nodes are properly canonicalized.

janz93 committed:
test(deps): update OpenSSL API usage for Ruby 3.2+ compatibility

Ruby 3.2 deprecated the old OpenSSL::Digest::SHA256 and OpenSSL::Cipher::AES
constructors. Tests now use the unified factory methods to avoid deprecation
warnings and ensure compatibility with future Ruby versions.

janz93 committed:
test: verify bank certificates survive encrypted serialization round-trip

HPB-initialized certificates must remain functional after keyring dump/reload
to prevent re-initialization on every application start. These tests catch
regressions where certificates are silently dropped during encryption.

janz93 committed:
feat: support H005 (EBICS 3.0) response format for HAA and HTD order types

Banks running EBICS H005 changed their response structure from simple
OrderTypes strings to structured Service elements. The gem now handles
both formats based on the negotiated protocol version, enabling
compatibility with H005-enabled banking platforms.

janz93 committed:
### Fix 1 (Critical): Add `Scope` to CCT service parameters

**File:** `epic-3/lib/epics/factories/request_factory/v3.rb:186-188`

Add `scope: 'DE'` (or make it configurable) to the `create_cct` method so the BTU
request includes `<Scope>DE</Scope>` in the `<Service>` element.

**However**, the correct Scope value is **bank-specific**. The ideal solution is to
read the bank's HTD response and use the service definitions it reports. See Fix 4.

### Fix 2 (Critical): Correct CDD Scope for Sparkasse

**File:** `epic-3/lib/epics/factories/request_factory/v3.rb:190-192`

Change `scope: 'GLB'` to the value the Sparkasse bank expects (likely `DE`), or
make it configurable per-bank. Again, the ideal solution is HTD-driven service
discovery.

### Fix 3 (Important): Add `<Container>` support to BTU builder

**File:** `epic-3/lib/epics/builders/order_details_builder/v3.rb:36-53`

Add a `container_type:` parameter to `add_btu_order_params` and generate
`<Container containerType="..."/>` the same way `add_btd_order_params` does.
Then propagate this parameter through the factory methods.

janz93 committed:
allow service_option: 'VOO'

janz93 committed:
Overwrites for c53

janz93 committed:
Enrich options for ebics 3

Nymuxyzo commented:
Starting in November 2026, EBICS 3.0 will essentially become mandatory. What steps still need to be taken, and how can we help?

janz93 commented:
@Nymuxyzo testing this would be a huge help.
Especially the with focus on the BTU to create transactions

janz93 commented:
@jplot I adjust the PR to fix issues [in a Fork](https://github.com/janz93/epics/commits/feat/ebics3-certificates/). Feel free to merge them into yours to bring to PR forward.

suzumejakku commented:
Hi,
I am currently using @janz93 fork he mentioned. The bank I am connecting to is Raiffeisen (Switzerland). The bank does not accept my C54 order if I don't set the scope to CH. But the code doesn't seem to let us set the scope ?

tobischo commented:
Alright, so here is our plan now:
1. we will merge this PR in the coming days, as not much is moving here right now and overall it looks fine. This will not immediately result in a release
2. Any obvious issues such as the problem mentioned by @suzumejakku and discussed in #183 will be applied in follow up PRs
3. We will publish a release candidate version which you can then use for testing and providing feedback
